src/metatrain/experimental/nativepet/modules/nef.py

```diff
@@ -0,0 +1,171 @@
+    # first, we shift the cell_shifts so that the minimum value is 0
+    min_cell_shift_x = cell_shifts_x.min()
+    cell_shifts_x = cell_shifts_x - min_cell_shift_x
+    negative_cell_shifts_x = negative_cell_shifts_x - min_cell_shift_x
+
+    min_cell_shift_y = cell_shifts_y.min()
+    cell_shifts_y = cell_shifts_y - min_cell_shift_y
+    negative_cell_shifts_y = negative_cell_shifts_y - min_cell_shift_y
+
+    min_cell_shift_z = cell_shifts_z.min()
+    cell_shifts_z = cell_shifts_z - min_cell_shift_z
+    negative_cell_shifts_z = negative_cell_shifts_z - min_cell_shift_z
+
+    max_centers_neigbors = centers.max() + 1  # same as neighbors.max() + 1
+    max_shift_x = cell_shifts_x.max() + 1
+    max_shift_y = cell_shifts_y.max() + 1
+    max_shift_z = cell_shifts_z.max() + 1
+
+    size_1 = max_shift_z
+    size_2 = max_shift_y * size_1
+    size_3 = max_shift_x * size_2
+    size_4 = max_centers_neigbors * size_3
+
+    unique_id = (
+        centers * size_4
+        + neighbors * size_3
+        + cell_shifts_x * size_2
+        + cell_shifts_y * size_1
+        + cell_shifts_z
+    )
+
+    # the inverse is the same, but centers and neighbors are swapped
+    # and we use the negative values of the cell_shifts
+    unique_id_inverse = (
+        neighbors * size_4
+        + centers * size_3
+        + negative_cell_shifts_x * size_2
+        + negative_cell_shifts_y * size_1
+        + negative_cell_shifts_z
+    )
+
+    unique_id_argsort = unique_id.argsort()
+    unique_id_inverse_argsort = unique_id_inverse.argsort()
+
+    corresponding_edges = torch.empty_like(centers)
+    corresponding_edges[unique_id_argsort] = unique_id_inverse_argsort
+
+    return corresponding_edges.to(array.dtype)
+
+
+def edge_array_to_nef(
+    edge_array,
+    nef_indices,
+    mask: Optional[torch.Tensor] = None,
+    fill_value: float = 0.0,
+):
+    # converts an edge array to a NEF array
+    if mask is None:
+        return edge_array[nef_indices]
+    else:
+        return torch.where(
+            mask.reshape(mask.shape + (1,) * (len(edge_array.shape) - 1)),
+            edge_array[nef_indices],
+            fill_value,
+        )
+
+
+def nef_array_to_edges(nef_array, centers, nef_to_edges_neighbor):
+    # converts a NEF array to an edge array. Most often, this converts
+    # a NEF array (three-dimensional, where the dimensions are the nodes,
+    # the edges, and the features) to an edge array (two-dimensional,
+    # where the dimensions are the edges and the features).
+    return nef_array[centers, nef_to_edges_neighbor]
+
+
+def compute_neighbors_pos(
+    nef_indices: torch.Tensor,
+    corresponding_edges: torch.Tensor,
+    nef_mask: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Creates a reversed neighborlist, where for each
+    center atom `i` and its neighbor `j` in the original
+    neighborlist, the position of atom `i` in the list
+    of neighbors of atom `j` is returned.
+    """
+    N, K = nef_indices.shape
```

Comment:
What are N and K? We need either a comment, or, even better, informative variable names